File: src/dashboard/src/types/monitoring.ts
/**
 * Monitoring types for project health checks.
 */

// CheckType is now dynamic — any string check ID from project config
export type CheckType = string

export type CheckStatus = 'idle' | 'running' | 'success' | 'failure'

export interface CheckResult {
  project_id: string
  check_type: CheckType
  status: CheckStatus
  exit_code: number | null
  duration_ms: number | null
  stdout: string
  stderr: string
}

export interface ProjectHealth {
  project_id: string
  project_name: string
  project_path: string
  checks: Record<CheckType, CheckResult>
  last_updated: string
}

// SSE Event payloads
export interface CheckStartedPayload {
  project_id: string
  check_type: CheckType
  started_at: string
}

export interface CheckProgressPayload {
  project_id: string
  check_type: CheckType
  output: string
  is_stderr: boolean
}

export interface CheckCompletedPayload {
  project_id: string
  check_type: CheckType
  status: CheckStatus
  exit_code: number
  duration_ms: number
  stdout: string
  stderr: string
}

// Per-project health check config from API
export interface CheckConfigEntry {
  label: string
  command: string
}

export interface CheckConfig {
  project_id: string
  check_types: string[]  // ordered list of check type IDs from backend
  checks: Record<string, CheckConfigEntry>
}

// Workflow check types for monitor integration
export type WorkflowCheckStatus = 'idle' | 'running' | 'success' | 'failure'

export interface WorkflowCheck {
  id: string
  name: string
  description: string
  status: WorkflowCheckStatus
  lastRunAt: string | null
  lastRunDuration: number | null // seconds
}

// Vault Health JSON result (parsed from check stdout)
export interface VaultHealthCheck {
  name: string
  status: 'pass' | 'warn' | 'fail'
  count: number
  details: string[]
}

export interface VaultHealthMetrics {
  total_notes: number
  total_links: number
  broken_links: number
  orphan_notes: number
  missing_frontmatter: number
  broken_images: number
}

export interface VaultHealthResult {
  status: 'healthy' | 'degraded' | 'unhealthy'
  timestamp: string
  metrics: VaultHealthMetrics
  checks: VaultHealthCheck[]
}

// Project Context types
export interface DevDocFile {
  name: string
  path: string
  content: string
  modified_at: string
}

export interface SessionInfo {
  session_id: string
  tasks_count: number
  agents_count: number
  iteration_count: number
  current_task_id: string | null
}

export interface ProjectContext {
  project_id: string
  project_name: string
  project_path: string
  claude_md: string | null
  dev_docs: DevDocFile[]
  session_info: SessionInfo | null
}

File: src/backend/models/monitoring.py
"""Monitoring models for project health checks."""

from datetime import datetime

from utils.time import utcnow
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field


class CheckType(str, Enum):
    """Types of project checks."""

    TEST = "test"
    LINT = "lint"
    TYPECHECK = "typecheck"
    BUILD = "build"


class CheckStatus(str, Enum):
    """Status of a check."""

    IDLE = "idle"
    RUNNING = "running"
    SUCCESS = "success"
    FAILURE = "failure"


class CheckResult(BaseModel):
    """Result of a single check."""

    check_type: CheckType
    status: CheckStatus = CheckStatus.IDLE
    exit_code: int | None = None
    stdout: str = ""
    stderr: str = ""
    duration_ms: int | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None

    model_config = ConfigDict(use_enum_values=True)


class ProjectHealth(BaseModel):
    """Overall project health status."""

    project_id: str
    project_name: str
    project_path: str
    checks: dict[CheckType, CheckResult] = Field(default_factory=dict)
    last_updated: datetime = Field(default_factory=utcnow)

    def get_check(self, check_type: CheckType) -> CheckResult:
        """Get or create a check result."""
        if check_type not in self.checks:
            self.checks[check_type] = CheckResult(check_type=check_type)
        return self.checks[check_type]

    def update_check(self, result: CheckResult) -> None:
        """Update a check result."""
        self.checks[result.check_type] = result
        self.last_updated = utcnow()


class CheckStartedPayload(BaseModel):
    """Payload when a check starts."""

    project_id: str
    check_type: CheckType
    started_at: datetime


class CheckProgressPayload(BaseModel):
    """Payload for check progress (streaming output)."""

    project_id: str
    check_type: CheckType
    output: str
    is_stderr: bool = False


class CheckCompletedPayload(BaseModel):
    """Payload when a check completes."""

    project_id: str
    check_type: CheckType
    status: CheckStatus
    exit_code: int
    duration_ms: int
    stdout: str
    stderr: str
